New Resource - `azurerm_web_pubsub_custom_certificate`

xiaxyi opened the pull request:
acc test:
--- PASS: TestAccCustomCertBindingWebPubsub_basic (541.53s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/signalr       542.248s

mbfrahry commented on internal/services/signalr/web_pubsub_custom_certificate_binding.go:
```diff
@@ -0,0 +1,282 @@
+			}
+
+			if certVersion := keyVaultCertificateId.Version; certVersion != "" {
+				if webPubsubCustomCertBinding.CertificateVersion != "" && certVersion != webPubsubCustomCertBinding.CertificateVersion {
```
Can we have a test that hits this use case?

mbfrahry commented on internal/services/signalr/web_pubsub_custom_certificate_binding.go:
```diff
@@ -0,0 +1,282 @@
+			),
+		},
+
+		"certificate_version": {
```
And a test that uses this?

mbfrahry commented on internal/services/signalr/web_pubsub_custom_certificate_binding.go:
```diff
@@ -0,0 +1,282 @@
+				customCert.Properties.KeyVaultSecretVersion = utils.String(certVersion)
+			}
+
+			if _, err := client.CustomCertificatesCreateOrUpdate(ctx, id, customCert); err != nil {
```
Would `CustomCertificatesCreateOrUpdateThenPoll` work here instead of using `WaitForStateContext`?
```suggestion
			if _, err := client.CustomCertificatesCreateOrUpdate(ctx, id, customCert); err != nil {
```

mbfrahry commented on internal/services/signalr/web_pubsub_custom_certificate_binding.go:
```diff
@@ -0,0 +1,282 @@
+}
+
+func (r CustomCertBindingWebPubsubResource) ResourceType() string {
+	return "azurerm_web_pubsub_custom_certificate_binding"
```
From looking at the portal, this is just a certificate. 

<img width="1168" alt="image" src="https://user-images.githubusercontent.com/6515518/227655631-56f93a5e-aa37-422c-afba-4c9b73fd01d2.png">

Does this need to be named `azurerm_web_pubsub_custom_certificate_binding` or would `azurerm_web_pubsub_custom_certificate` be ok?

mbfrahry commented on internal/services/signalr/web_pubsub_custom_certificate_binding.go:
```diff
@@ -0,0 +1,282 @@
+				return fmt.Errorf("checking for presence of existing web pubsub service custom cert binding error %s: %+v", id, err)
+			}
+
+			if !response.WasNotFound(existing.HttpResponse) {
```
Can we add a requiresImport test as well?

mbfrahry reviewed (changes requested):
Hey @xiaxyi. Thanks for getting this resource written. It is mostly good but it could use a rename, a couple more tests, and some logic that can be removed in favor of a builtin function. I've detailed all that below. Let me know if you have any questions. Thanks!

ms-zhenhua commented on internal/services/signalr/web_pubsub_custom_certificate_binding.go:
```diff
@@ -0,0 +1,282 @@
+				if response.WasNotFound(resp.HttpResponse) {
+					return metadata.MarkAsGone(id)
+				}
+				return fmt.Errorf("reading web pubsub custom certificate binding %s: %+v", id, err)
```
```suggestion
				return fmt.Errorf("retrieving %s: %+v", *id, err)
```

ms-zhenhua commented on internal/services/signalr/web_pubsub_custom_certificate_binding.go:
```diff
@@ -0,0 +1,282 @@
+			state := CustomCertBindingWebPubsubModel{
+				Name:               id.CustomCertificateName,
+				CustomCertId:       certId,
+				WebPubsubId:        webPubsubId,
```
```suggestion
				WebPubsubId:        webpubsub.NewWebPubSubID(id.SubscriptionId, id.ResourceGroupName, id.WebPubSubName).ID(),
```

ms-zhenhua commented on internal/services/signalr/web_pubsub_custom_certificate_binding.go:
```diff
@@ -0,0 +1,282 @@
+				return err
+			}
+
+			var webPubsubCustomCertBinding CustomCertBindingWebPubsubModel
```
Suggest keep consistent with other resources
```suggestion
			var model CustomCertBindingWebPubsubModel
```

ms-zhenhua commented on internal/services/signalr/web_pubsub_custom_certificate_binding.go:
```diff
@@ -0,0 +1,282 @@
+			}
+			props := existing.Model
+			if props == nil {
+				return fmt.Errorf("retrieving %s: `model` was nil", *id)
```
```suggestion
				return fmt.Errorf("retrieving %s: props was nil", *id)
```

ms-zhenhua commented on internal/services/signalr/web_pubsub_custom_certificate_binding_resource_test.go:
```diff
@@ -0,0 +1,146 @@
+		}
+		return nil, fmt.Errorf("retrieving %s: %+v", *id, err)
+	}
+	return utils.Bool(true), nil
```
```suggestion
	return utils.Bool(resp.Model != nil), nil
```

ms-zhenhua commented on website/docs/r/web_pubsub_custom_certificate_binding.html.markdown:
```diff
@@ -0,0 +1,129 @@
+
+The following arguments are supported:
+
+* `name` - (Required) The name of the Web Pubsub Custom Certificate Binding. Changing this forces a new resource to be created.
```
```suggestion
* `name` - (Required) The name of the Web Pubsub Custom Certificate Binding. Changing this forces a new Azure Web Pubsub Custom Certificate to be created.
```

xiaxyi commented:
Thanks @mbfrahry for the review. I changed the cert_version to totally computed one as it can be retrieved from the cert id. Also, removed the binding from the naming.

acc test results:
```
--- PASS: TestAccCustomCertWebPubsub_basic (557.53s)
--- PASS: TestAccCustomCertWebPubsub_requiresImport (557.89s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/signalr       558.695s


```

mbfrahry reviewed (approved):
LGTM! Thanks for this @xiaxyi!
